File: lintcode/valid-parentheses.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/valid-parentheses/

class Solution:
    # @param {string} s A string
    # @return {boolean} whether the string is a valid parentheses
    def isValidParentheses(self, s):
        # Write your code here
        stack = []
        for i in s:
            if i == '(' or i == '[' or i == '{':
                stack.append(i)
            else:
                if not stack:
                    return False
                if (stack[-1] == '(' and i != ')') or (stack[-1] == '[' and i != ']') or (stack[-1] == '{' and i != '}'):
                    return False
                stack.pop()
        if stack:
            return False
        return True

File: misc/google/valid-parentheses.cc
// Question: 一个由()组成的字符串，判断是否是valid。

bool valid(const string& p) {
    int d = 0;
    for (char c : p) {
        if (c == '(') {
            ++d;
        } else {
            --d;
            if (d < 0) {
                return false;
            }
        }
    }
    return true;
}
